### PR TITLE
UGENE-7375. Check user forward and reverse primers

### PR DESCRIPTION
**Pull request by evelinabis**
Added user primers to report.

### CODE REVIEW

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 1]
```diff
@@ -544,30 +629,113 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+        report += getReversePrimerReportResult(backboneSequence, sequence.mid(reverse.startPos, reverse.length));
     }
     return report;
 }
 
+QString PCRPrimerDesignForDNAAssemblyTask::getPairReportForUserPrimers() const {
+    QString report_;
+    if (!settings.forwardUserPrimer.isEmpty() && !settings.reverseUserPrimer.isEmpty()) {
+        report_ += getPrimerReportTitle("C", true);
+        report_ += getForwardPrimerReportResult(backboneSequence, settings.forwardUserPrimer);
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp]: Maybe I didn't mention it, but the backbone sequence shouldn't be added to user primers

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 2]
```diff
@@ -544,30 +629,113 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+}
+
+QString PCRPrimerDesignForDNAAssemblyTask::getUserPrimersUnwantedConnectionsReport() const {
+    using Seq = SeqToSearchInThem;
+    const QList<Seq> allTypes = { Seq::ForwardUser, Seq::ReverseUser, Seq::Sequence, Seq::RevComplSeq, Seq::OtherSeq };
+    const QMap<Seq, QString> toString = { {Seq::ForwardUser, tr("User forward primer")},
+                                          {Seq::ReverseUser, tr("User reverse primer")},
+                                          {Seq::Sequence, tr("Sequence")},
+                                          {Seq::RevComplSeq, tr("Reverse complement sequence")},
+                                          {Seq::OtherSeq, tr("Other sequences")} };
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp]: `Other sequences in PCR reaction`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 3]
```diff
@@ -544,30 +629,113 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+    // connections have been formed.
+    QString report_ = tr("<br><h3>Unwanted connections user primers</h3>");
+    QString table = "<table border=\"1\" cellpadding=\"4\">"
+                      "<tr>"
+                        "<td></td>"
+                        "<th>User forward</th>"
+                        "<th>User reverse</th>"
+                        "<th>Sequence</th>"
+                        "<th>Rev.compl.sequence</th>"
+                        "<th>Other sequences</th>"
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp]: `Other sequences in PCR reaction`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 4]
```diff
@@ -544,30 +629,113 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+    const QList<Seq> allTypes = { Seq::ForwardUser, Seq::ReverseUser, Seq::Sequence, Seq::RevComplSeq, Seq::OtherSeq };
+    const QMap<Seq, QString> toString = { {Seq::ForwardUser, tr("User forward primer")},
+                                          {Seq::ReverseUser, tr("User reverse primer")},
+                                          {Seq::Sequence, tr("Sequence")},
+                                          {Seq::RevComplSeq, tr("Reverse complement sequence")},
+                                          {Seq::OtherSeq, tr("Other sequences")} };
+
+    // Summary table for unwanted connections. Rows -- user primers, columns -- sequences with which unwanted
+    // connections have been formed.
+    QString report_ = tr("<br><h3>Unwanted connections user primers</h3>");
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp]: `Unwanted connections of user primers`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 5]
```diff
@@ -544,30 +629,113 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+        report_ += getReversePrimerReportResult(backboneSequence, settings.reverseUserPrimer);
+    }
+    return report_;
+}
+
+QString PCRPrimerDesignForDNAAssemblyTask::getUserPrimersUnwantedConnectionsReport() const {
+    using Seq = SeqToSearchInThem;
+    const QList<Seq> allTypes = { Seq::ForwardUser, Seq::ReverseUser, Seq::Sequence, Seq::RevComplSeq, Seq::OtherSeq };
+    const QMap<Seq, QString> toString = { {Seq::ForwardUser, tr("User forward primer")},
+                                          {Seq::ReverseUser, tr("User reverse primer")},
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp]: `Forward user primer` and `Reverse user primer`

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 6]
```diff
@@ -544,30 +629,113 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+                                          {Seq::RevComplSeq, tr("Reverse complement sequence")},
+                                          {Seq::OtherSeq, tr("Other sequences")} };
+
+    // Summary table for unwanted connections. Rows -- user primers, columns -- sequences with which unwanted
+    // connections have been formed.
+    QString report_ = tr("<br><h3>Unwanted connections user primers</h3>");
+    QString table = "<table border=\"1\" cellpadding=\"4\">"
+                      "<tr>"
+                        "<td></td>"
+                        "<th>User forward</th>"
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp]: The same as [here](https://github.com/ugeneunipro/ugene/pull/525/files#r690046092)

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]
Some general comments because I couldn't find the certain lines where they should be placed:
1. I don't have any comments against the general table you have made (with + and -) despite the fact it wasn't described in the issue, but you should add a legend or a description.
2. You shouldn't search hetero-dimers between the same sequences. I mean, that you have searched hetero-dimers between "User forward primer" and "User forward primer" (the same about reverse). 
image
3. Also make it ""Forward user" and "Reverse user" for both lines
image

**Comment by DmitriiSukhomlinov** [comment]
> Some general comments because I couldn't find the certain lines where they should be placed:
> 
> 1. I don't have any comments against the general table you have made (with + and -) despite the fact it wasn't described in the issue, but you should add a legend or a description.
> 2. You shouldn't search hetero-dimers between the same sequences. I mean, that you have searched hetero-dimers between "User forward primer" and "User forward primer" (the same about reverse).
>    image
> 3. Also make it ""Forward user" and "Reverse user" for both lines
>    image

2)  isn't fixed

File: src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp [hunk 7]
```diff
@@ -544,30 +637,115 @@ QString PCRPrimerDesignForDNAAssemblyTask::getPairReport(U2Region forward,
...
+    return report_;
+}
+
+QString PCRPrimerDesignForDNAAssemblyTask::getUserPrimersUnwantedConnectionsReport() const {
+    using Seq = SeqToSearchInThem;
+    const QList<Seq> allTypes = { Seq::ForwardUser, Seq::ReverseUser, Seq::Sequence, Seq::RevComplSeq, Seq::OtherSeq };
+    const QMap<Seq, QString> toString = { {Seq::ForwardUser, tr("Forward user primer")},
+                                          {Seq::ReverseUser, tr("Reverse user primer")},
+                                          {Seq::Sequence, tr("Sequence")},
+                                          {Seq::RevComplSeq, tr("Reverse complement sequence")},
```
Comment (DmitriiSukhomlinov) [inline src/plugins/pcr_primer_design_for_dna_assembly/src/tasks/PCRPrimerDesignForDNAAssemblyTask.cpp:585]: `Reverse complementary sequence`

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]
2) and a  minor comment

**Review by smlnkcv — APPROVED** [review]

**Comment by evelinabis** [comment]
#538

**Pull request closed without merging**